Branchprotector and Tide: Use GetPresubmits

alvaroaleman opened the pull request:
Another step towards fixing #13370, this PR changes the branchprotector and Tide to use `GetPresubmits` instead of directly accessing the `Presubmits` property

alvaroaleman commented on prow/config/branch_protection.go:
```diff
@@ -241,15 +249,28 @@ func (c *Config) GetBranchProtection(org, repo, branch string) (*Policy, error)
+	// whenever it triggers Jobs for a PR that has branch as base. Consequently either:
+	// * The request is already cached because Trigger did it before, we dont have to pay a token
+	// * We do the request and it will get cached, making Trigger not have to pay a token
+	sha, err := gitHubClient.GetRef(org, repo, "heads/"+branch)
```
...However on the first run of the branchprotector after this change, it may do some more `GetRef` calls, depending on how many branches no one ever used as base for a PR exist

alvaroaleman replied:
Using `git ls-remote` is a great idea! I wouldn't want to use that by default because I am too afraid its result will differ from what we get from the github API but for this particular case here we can certainly use it.

alvaroaleman replied:
On second thought though, this will impact the runtime quite a bit. When playing around, a `git ls-remote` takes 1-1.5S, if we do this once for every branch that may take a moment. WDYT?

alvaroaleman commented:
/assign @stevekuznetsov @cjwagner

cjwagner commented on prow/tide/tide.go:
```diff
@@ -1151,7 +1152,16 @@ func (c *Controller) presubmitsByPull(sp *subpool) (map[int][]config.Presubmit,
+	for _, pr := range sp.prs {
+		headRefs = append(headRefs, string(pr.HeadRefOID))
+	}
+	configPresubmits, err := c.config().GetPresubmits(c.gc, sp.org+"/"+sp.repo, sp.sha, headRefs...)
```
This does not produce the correct set of presubmits. This assumes that all PRs in a subpool are part of a single batch and won't be merged individually or in a smaller batch.

The presubmits that are required for a set of PRs now need to be determined for every set of PRs we consider, a simple union is no longer sufficient.

alvaroaleman replied:
Yeah you are right, I will update this

cjwagner commented on prow/config/tide.go:
```diff
@@ -470,22 +471,35 @@ func parseTideContextPolicyOptions(org, repo, branch string, options TideContext
+	if err != nil {
+		return nil, fmt.Errorf("failed to get SHA for branch %q: %v", branch, err)
+	}
+	presubmits, err := c.GetPresubmits(gitClient, org+"/"+repo, sha)
```
This ignores the changes in the PRs by just using the base branch commit.

alvaroaleman commented on prow/cmd/branchprotector/protect.go:
```diff
@@ -182,7 +192,10 @@ func (p *protector) protect() {
+	// Using PresubmitsStatic here is safe because this is only about getting to
+	// know which repos exist. Repos that use in-repo config will appear here,
+	// because we generate a verification job for them
+	for repo := range p.cfg.PresubmitsStatic() {
```
My understanding from what we discussed in the sig-testing meeting a couple of weeks back was that we will have the branchprotector use the config that is on the branch its supposed to protect and that we will just create a green no-op status check on Pulls that remove required presubmits, as that is pretty easy to figure out.

Did I understand that wrong?

alvaroaleman commented:
@cjwagner @stevekuznetsov updated this to use closures as well, hence it wont consume any additional tokens unless `inrepoconfig` is enabled. PTAL.

stevekuznetsov commented on prow/config/branch_protection.go:
```diff
@@ -241,15 +249,23 @@ func (c *Config) GetBranchProtection(org, repo, branch string) (*Policy, error)
+	refGetter := func() (string, error) {
+		return gitHubClient.GetRef(org, repo, "heads/"+branch)
+	}
+	presubmits, err := c.GetPresubmits(gitClient, org+"/"+repo, refGetter)
```
Wait so if we're only protecting the centrally-configured jobs and the synthetic job that's generated, why can't we use the static list here?

alvaroaleman replied:
So my understanding from what we discussed a couple of weeks back was that Cole suggested that we just have the Branchprotector get the jobs from the branch it should protect.

You both seem to not remember that so maybe my memories are wrong.

I personally have no strong opinions about this, but I don't see a technical reason why that shouldn't be possible. `inrepoconfig` must create dummy statuses if someone removes a context but other than that this should just work.

stevekuznetsov replied:
Oh, fetching from the branch itself -- that sounds good but you are going to need to run this protector really often to make it consistent.

alvaroaleman replied:
Yep, true. So as mentioned, I personally don't really care. Should I keep it or should I remove it and we consider this a problem for another day?

cjwagner replied:
We've tried to move away from posting "Skipped" statuses so I favor avoiding that and relying on Tide to enforce that all the required jobs have run.  It seems wrong to apply branch protection policies across an entire repo when we know that some of the requirements won't actually apply to some of the PRs and that there will also be requirements missing for some PRs.

alvaroaleman replied:
Removed it

alvaroaleman commented:
@cjwagner @stevekuznetsov Would be great if you find the time to have another look :)

cjwagner commented on prow/tide/tide.go:
```diff
@@ -1163,12 +1174,17 @@ func (c *Controller) presubmitsByPull(sp *subpool) (map[int][]config.Presubmit,
-		if !ps.ContextRequired() {
-			continue
+	for _, pr := range sp.prs {
+		presubmits, err := c.config().GetPresubmits(c.gc, sp.org+"/"+sp.repo, refGetterFactory(sp.sha), refGetterFactory(string(pr.HeadRefOID)))
```
Why does `GetPresubmits` use getter functions for the shas? When do we call `GetPresubmits` without already having the SHAs available?

alvaroaleman replied:
Never in Tide but in all other current users for `GetPresubmits`:
* The `skip` plugin
* The `trigger` plugin
* In ` Deck

cjwagner commented on prow/tide/tide.go:
```diff
@@ -1153,6 +1158,12 @@ func (c *changedFilesAgent) prune() {
+}
+
 func (c *Controller) presubmitsByPull(sp *subpool) (map[int][]config.Presubmit, error) {
 	presubmits := make(map[int][]config.Presubmit, len(sp.prs))
```
As mentioned [here](https://github.com/kubernetes/test-infra/pull/13644#discussion_r310824633) this won't be sufficient for batches. We can't just union the presubmits required by each PR in the batch which is what we are doing now since this PR doesn't update `accumulateBatch` or `trigger`.

cjwagner commented on prow/tide/tide_test.go:
```diff
@@ -636,10 +637,15 @@ func TestDividePool(t *testing.T) {
 		},
 	}
```
Please add some test cases to validate the new behavior.

alvaroaleman commented:
@stevekuznetsov @cjwagner PTAL, I've updated the PR:

* The branchprotector does not use `inrepoconfig` anymore
* `Tide` no re-resolves the Presubmits for all unique baseSHA+headSHA(s) combinations:
    *  `presubmitsByPull`
    *  `acummulateBatch`
    * `pickBatch`
* Added tests for all of that. In order to be able to test the `Presubmit scoped to PR` cases I added  a package-level variable `FakeInRepoConfig` to the `config` package. Once the whole thing is implemented, I'll probably just replace it with a localgit-based helper

Unfortuantely this blew up the PR quite a bit, especially the tests.

stevekuznetsov commented on prow/tide/tide.go:
```diff
@@ -1022,26 +1029,24 @@ func (c *Controller) trigger(sp subpool, presubmits map[int][]config.Presubmit,
-				return fmt.Errorf("failed to create a ProwJob for job: %q, PRs: %v: %v", spec.Job, prNumbers(prs), err)
-			}
-			c.logger.WithField("duration", time.Since(start).String()).Debug("Created ProwJob on the cluster.")
+	for _, ps := range presubmits {
```
So technically the previous implementation was not vulnerable to triggering things in `presubmits` that didn't refer to any `pr` -- this impl triggers all presubmits. Do we care? Why bother changing?

alvaroaleman replied:
So we gets changed conceptionally is that now each baseSHA+headSHA(s) combination may have a unique set of presubmits. The result is that we can not view them as decoupled anymore, which is why I essentialy moved the responsibility of picking the right presubmits out of the `trigger` function and then made the `presubmits` arg a slice to properly reflect that.

stevekuznetsov commented on prow/tide/tide_test.go:
```diff
@@ -216,38 +228,60 @@ func TestAccumulateBatch(t *testing.T) {
+		t.Run(test.name, func(t *testing.T) {
+			config.FakeInRepoConfig = test.fakeInRepoConfig
+			defer func() {
+				config.FakeInRepoConfig = nil
```
Is this threadsafe in the context of parallel tests? Why do we need to mutate global state for mocking? Why not inject the dependency?

alvaroaleman replied:
Its not, which is not a problem as none of these tests are being executed in parallel. I can make `FakeInRepo` config a property of `config.Config` instead of a package-level var though.

cjwagner replied:
I think making it a file of the config struct instead of global would be nice. We can use the field annotation ``` `json:"-"` ``` to prevent users from specifying this in config since it is only for testing.

alvaroaleman replied:
Updated this

stevekuznetsov commented on prow/tide/tide_test.go:
```diff
@@ -216,38 +228,60 @@ func TestAccumulateBatch(t *testing.T) {
-					Number: pr.number,
-					SHA:    pr.sha,
-				})
+			merges, pending := c.accumulateBatch(subpool{org: "org", repo: "repo", prs: pulls, pjs: pjs, log: logrus.NewEntry(logrus.New())})
```
Might as well do `logrus.WithField` with the test name here

stevekuznetsov commented on prow/tide/tide.go:
```diff
@@ -1286,7 +1349,7 @@ type subpool struct {
+	cc map[int]contextChecker
 	// presubmit contains all required presubmits for each PR
 	// in this subpool
 	presubmits map[int][]config.Presubmit
```
Why does this still exist and who uses it, if we're dynamically resolving the list in `accumulateBatch`?

alvaroaleman replied:
It is being used in:

* `filterPR` to filter out PRs with failing contexts or contexts that are not controlled by Prow
* `accumulate` to determine the current state of PRs

cjwagner commented on prow/config/tide.go:
```diff
@@ -485,25 +486,42 @@ func parseTideContextPolicyOptions(org, repo, branch string, options TideContext
 	optional := sets.NewString(options.OptionalContexts...)
 
+	baseSHAGetter := func() (string, error) {
+		return gitHubClient.GetRef(org, repo, "heads/"+branch)
```
Actually it looks like the status controller doesn't have the baseSHA. I'd be happy with the status controller looking up the baseSHA separately (although only once per pool per sync loop) since we won't have looked up the baseSHA for any repos where no PRs match the tide query and because the status controller is considered best effort.
We probably still want the sync controller to use a consistent baseSHA though?

alvaroaleman replied:
Yeah, the consistency is a good point here. I have updated the code to pass it on from the sync controller/use a closure in the status controller

cjwagner commented on prow/tide/tide.go:
```diff
@@ -1145,6 +1150,22 @@ func (c *changedFilesAgent) prChanges(pr *PullRequest) config.ChangedFilesProvid
+				return nil, err
+			}
+
+			result = append(result, changes...)
```
nit: If multiple PRs touch the same file this function will return duplicate entries in the slice. I don't think duplicates will cause any unexpected behavior, but it would probably be cleaner to use a set here.

cjwagner commented on prow/tide/tide.go:
```diff
@@ -1163,12 +1190,17 @@ func (c *Controller) presubmitsByPull(sp *subpool) (map[int][]config.Presubmit,
-		if !ps.ContextRequired() {
-			continue
+	for _, pr := range sp.prs {
+		presubmits, err := c.config().GetPresubmits(c.gc, sp.org+"/"+sp.repo, refGetterFactory(sp.sha), refGetterFactory(string(pr.HeadRefOID)))
```
nit: `presubmits` is shadowed here. It looks like this still works properly because the `record` closure captures the correct `presubmits` variable, but it would be a clearer to use a different variable name.

cjwagner commented on prow/tide/status.go:
```diff
@@ -288,10 +290,16 @@ func (sc *statusController) setStatuses(all []PullRequest, pool map[string]PullR
+		branch := string(pr.BaseRef.Name)
+		headSHA := string(pr.HeadRefOID)
+		baseSHAGetter := func() (string, error) {
+			return sc.ghc.GetRef(org, repo, "heads/"+branch)
```
Well want to cache the sha so that we only look it up once per pool instead of once per PR. This doesn't block merge though since we won't actually call the baseSHAGetter at all right now.
